<!doctype html>
<html lang="en">

    <!-- meta data and javascript import -->
	<head>
		<meta charset="utf-8">
		<title>Whole Exome Sequencing for Clinical Diagnostics</title>
		<link rel="stylesheet" href="css/reveal.css">
		<link rel="stylesheet" href="css/base.css">
		<link rel="stylesheet" href="css/theme/black.css" id="theme">
		<!-- Code syntax highlighting -->
		<link rel="stylesheet" href="lib/css/zenburn.css">
		<!-- Make sure jquery is loaded before highcharts -->
		<!-- Jquery not required for Highstock charts which has lightweight jquery embeded -->
    	<!-- <script type="text/javascript" src="js/jquery-1.11.3.min.js"></script> -->
    	<!-- <script type="text/javascript" src="js/highcharts.js"></script> -->
		<!-- Import data.js to allow importing of csv files -->
		<!-- had issues with cross origin requests when accessing data stored locally -->
		<!-- Not currently implemmented in this presentation -->
		<script type="text/javascript" src="modules/data.js"></script>
		<!-- Creates dark themed plots throughout the presentation -->
		<script src="js/themes/gray.js"></script>
	</head>

    <!-- body content -->
    <body>

		<div class="reveal">

			<header class='headings'>
				<span></span>
			</header>

			<!-- Any section element inside of this container is displayed as a slide -->
			<div class="slides">
                <!-- title slide -->
  				      <section>
					          <h3>Whole Exome Sequencing for Clinical Diagnostics</h3>
					          <p style="font-size: 20px">
						        <!-- <small> -->
						        <a href="http://sirselim.github.io/" target="blank">Miles Benton</a>
                    <!-- </small> -->
                    </p>
                    <p style="font-size: 14px">
                    School of Medical Sciences, Institute of Health and Biomedical Innovation, Queensland University of Technology, Brisbane, Australia <br>
                    </p>
                    <p style="font-size: 16px">
                      6<sup style="font-size: 10px">th</sup> December 2016
					          </p>
                  <center>
                    <img src="images/ihbi_building.jpg" width=40% />
                  </center>
                </section>

                <!-- Norfolk Island  -->
                <!--- Vertical slides take affect here following double section breaks -->
                <section>
                  <section>
                    <h2>Norfolk Island</h2>
                  </section>
                  <section>
                    <img src="images/NI_map.png" width=60% />
                    <p style="font-size: 20px"><a href="https://github.com/gringer/bioinfscripts/blob/master/perlshaper.pl" target="blank">https://github.com/gringer/bioinfscripts/blob/master/perlshaper.pl</a></p>
                  </section>
                  <section>
                    <img src="images/NorfolkIsland.png">
                  </section>
                  <section>
                    <img src="images/NIpedigree.png"><br />
                    ~6000 member pedigree<br />
                    <p style="font-size: 14px">Macgregor S <i>et al.,:</i> <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2987173/" target="blank"><i>Legacy of mutiny on the Bounty: founder effect and admixture on Norfolk Island.</i></a> Eur J Hum Genet. 2010; 18: 67–72.</p>
                  </section>
                  <section>
                	<h3>Mitochondrial Ancestry</h3>
                    <img src="images/NI_ped_mt.png"><br />
                      40% of current population haplogroup <span style="color:lightblue">B4a1a[...]</span><br />

                      <p style="font-size: 14px">Benton MC <i>et al.,:</i> <a href="https://investigativegenetics.biomedcentral.com/articles/10.1186/s13323-015-0028-9" target="blank"><i>“Mutiny on the Bounty”: the genetic history of Norfolk Island reveals extreme gender-biased admixture.</i></a> Investigative Genetics 2015, 6:11.</p>
                  </section>
                </section>

                <!-- WES pipeline  -->
                <!--- Vertical slides take affect here following double section breaks -->
                <section>
                  <section>
                    <h2>Diagnostics: Whole Exome Sequencing </h2>
                  </section>
                  <section>
                    <h3>Ion Proton</h3>
                      <img src="images/ionsequencers.png" width=40%>
                      <img src="images/pI_chip.jpg" width=23.3%>
                     <ul>
                      <li> 2 Human Exomes </li>
                      <li> 60-80M filtered reads </li>
                      <li> up to 200bp reads </li>
                      <li> 2-4 hours run </li>
                    </ul>
                      <p style='text-align: center; font-size: xx-small;'>&copy; 2016 Thermo Fisher Scientific Inc.</p>
                  </section>
                  <section>
                    <p style="font-size: largest;">After QC, alignment and variant calling get between 28-42K variants per exome</p><br />
                    <p style="font-size: largest;">Diagnostics used to take ~4 weeks from generating the data to filtering down the variants and writing a report for the clinician</p><br />
                    <p style="font-size: largest;">We needed to make this process faster...</p>
                  </section>
                  <section>
                    <h3>Custom pipeline...</h3>
                     <ul>
                      <li> Sequencing, QC and alignment </li>
                      <li> annotation (VEP, SnpSift, dbNSFP, dbSNP) </li>
                      <li> variant filtering (bash and R) </li>
                      <li> report generation (RMarkdown) </li>
                    </ul>
                  </section>
                  <section>
                    <img src="images/taurus_WES_pipeline.png">    
                  </section>
                </section>

                <!-- WES Report -->
                <!--- Vertical slides take affect here following double section breaks -->
                <section>
                <h2>Automated Reporting</h2>
                </section>

                <!-- WES: interactive filtering and validation -->
                <!--- Vertical slides take affect here following double section breaks -->
                <section>
                <h2>WES: interactive filtering and validation</h2>
                </section>

            </div>
        </div>

    <script type="text/javascript" src="lib/js/head.min.js"></script>
    <script type="text/javascript" src="js/reveal.js"></script>
		<script type="text/javascript" src="js/d3.min.js"></script>
		<script type="text/javascript" src="js/barchart.js"></script>

        <script>
            // Full list of configuration options available at:
            // https://github.com/hakimel/reveal.js#configuration
            Reveal.initialize({
                controls: true,
                progress: true,
                history: true,
                center: true,
                keyboard: true,
                transition: 'none', // none/fade/slide/convex/concave/zoom
                // Optional reveal.js plugins
                dependencies: [
                    { src: 'lib/js/classList.js', condition: function() { return !document.body.classList; } },
                    { src: 'plugin/markdown/marked.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
                    { src: 'plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
                    { src: 'plugin/highlight/highlight.js', async: true, condition: function() { return !!document.querySelector( 'pre code' ); }, callback: function() { hljs.initHighlightingOnLoad(); } },
                    { src: 'plugin/zoom-js/zoom.js', async: true },
                    { src: 'plugin/notes/notes.js', async: true }
                ]
            });
        </script>

    </body>
</html>